src/Restorer.ts

```diff
@@ -77,9 +75,65 @@ class Restorer extends TabsterPart<RestorerProps> implements RestorerInterface {
     };
 }
 
+class History {
+    private static readonly DEPTH = 10;
+    private _stack: WeakHTMLElement<HTMLElement>[] = [];
+    private _getWindow: GetWindow;
+    constructor(getWindow: GetWindow) {
+        this._getWindow = getWindow;
+    }
+    /**
+     * Push a weak element to the top of the history stack.
+     * If the stack is full, the bottom weak element is removed.
+     * If the element is already at the top of the stack, it is not duplicated.
+     */
+    push(element: HTMLElement): void {
+        // Don't duplicate the top of history
+        if (this._stack[this._stack.length - 1]?.get() === element) {
+            return;
+        }
+
+        if (this._stack.length > History.DEPTH) {
+            this._stack.shift();
+        }
+        this._stack.push(
+            new WeakHTMLElement<HTMLElement>(this._getWindow, element)
+        );
+    }
+    /**
+     * Pop the first element from the history that satisfies the callback.
+     * The history is searched from the top to the bottom (from the most recent to the least recent).
+     *
+     * If a weak reference to the element is broken,
+     * or the element is no longer in the DOM,
+     * the element is removed from the top of the stack while popping.
+     *
+     * If no matching element is found, undefined is returned.
+     * If the stack is empty, undefined is returned.
+     */
+    pop(
+        filter: (element: HTMLElement) => boolean = () => true
+    ): HTMLElement | undefined {
+        const doc = this._getWindow().document;
+        for (let index = this._stack.length - 1; index >= 0; index--) {
+            const maybeElement = this._stack[index].get();
+            if (
+                !maybeElement ||
+                !dom.nodeContains(doc.body, dom.getParentElement(maybeElement))
+            ) {
+                this._stack.pop();
+            } else if (filter(maybeElement)) {
+                this._stack.splice(index, 1);
```

Comment:
I though you would have just continued popping, why do you think we should retain other history?

Reply:
Valid point, I was thinking it might be useful to not lose the restorer history while looking for a specific id. I guess we can just pop until we find the right one for now and then if this becomes necessary we can just introduce retention later on